### PR TITLE
[wip] Hybrid DataParallel/DDP

### PR DESCRIPTION
**Pull request by stephenroller**
Work in progress, you can ignore for now.

DistributedDataParallel has a mode where it acts like a DataParallel locally, but still broadcast gradients across all workers. This allows for a sort of "hybrid" training, where a local machine can use all GPUs for `-cands batch` negative samples, but still get the speedup from using multiple workers. Also can be run locally on one machine where it just behaves like DataParallel, getting rid of the need for having both code paths.

This is a "simpler" solution to the idea proposed by #1416. I suspect it will work well since individuals have reported there can be a sweet spot with number of negative candidates.

Issues:
- [x] DDP in this mode is broken internal fairseq env 20190211 (hard crashes). [cc @myleott] Fixed in torch stable (1.0.1post2).
- [ ] Cannot recover from a CUDA OOM. Need to file a bug with pytorch. For some reason, catching the RuntimeError doesn't clear any CUDA memory. This results in the dummy pass also OOMing, making the error unrecoverable. Frustratingly, this happens even when using only a world size of 1.
- [x] Make a minimum test case. File a pytorch bug.

### CODE REVIEW

**Comment by stephenroller** [comment]
Ugh, looks like you can't mark PRs is draft once you accidentally make them non-draft.

**Comment by stephenroller** [comment]
Tracking a larger bug in pytorch/pytorch#18853
